File: fix_utf8.py
import os

def fix_mojibake(filepath):
    try:
        with open(filepath, 'r', encoding='utf-8') as f:
            content = f.read()
    except Exception:
        return
        
    original = content
    # First, standard utf8 double-encoded mappings we know from the files
    reps = {
        'á': 'á',
        'é': 'é',
        'í\xad': 'í',
        'ó': 'ó',
        'ú': 'ú',
        'ñ': 'ñ',
        'Ñ': 'Ñ',
        'Ó': 'Ó',
        'í\x81': 'Á',
        'í\x89': 'É',
        'í\x8d': 'Í',
        'í\x9a': 'Ú',
        '·': '·',
        '—': '—',
        '¡': '¡',
        '›': '›',
        '¿': '¿',
    }
    
    for k, v in reps.items():
        content = content.replace(k, v)
        
    # After these replacements, any standalone í is likely 'í' where the soft hyphen got dropped
    # Let's do a safe replacement of í followed by common letters in those words
    content = content.replace('título', 'título')
    content = content.replace('Título', 'Título')
    content = content.replace('política', 'política')
    content = content.replace('Política', 'Política')
    content = content.replace('día', 'día')
    content = content.replace('Día', 'Día')
    content = content.replace('guía', 'guía')
    content = content.replace('Guía', 'Guía')
    content = content.replace('vacío', 'vacío')
    content = content.replace('río', 'río')
    content = content.replace('mínimo', 'mínimo')
    content = content.replace('límite', 'límite')
    content = content.replace('Límite', 'Límite')
    
    # Finally, just replace any remaining í with í as a catch-all since no valid word uses isolated í in Spanish
    content = content.replace('í', 'í')
    
    if content != original:
        with open(filepath, 'w', encoding='utf-8') as f:
            f.write(content)
        print(f"Fixed: {filepath}")

if __name__ == '__main__':
    for root, dirs, files in os.walk('.'):
        if '.git' in root or '.dart_tool' in root or 'build' in root or 'node_modules' in root or '.venv' in root or '__pycache__' in root:
            continue
        for file in files:
            if file.endswith('.dart') or file.endswith('.py') or file.endswith('.md'):
                fix_mojibake(os.path.join(root, file))
    print("Done fixing mojibake.")

File: fix_utf8.dart
import 'dart:io';

void fixMojibake(File file) {
  try {
    final original = file.readAsStringSync();
    var content = original;
    
    final reps = {
      'á': 'á',
      'é': 'é',
      'í\xad': 'í',
      'ó': 'ó',
      'ú': 'ú',
      'ñ': 'ñ',
      'Ñ': 'Ñ',
      'Ó': 'Ó',
      'í\x81': 'Á',
      'í\x89': 'É',
      'í\x8d': 'Í',
      'í\x9a': 'Ú',
      '·': '·',
      '—': '—',
      '¡': '¡',
      '›': '›',
      '¿': '¿',
    };
    
    reps.forEach((k, v) {
      content = content.replaceAll(k, v);
    });
    
    content = content.replaceAll('título', 'título');
    content = content.replaceAll('Título', 'Título');
    content = content.replaceAll('política', 'política');
    content = content.replaceAll('Política', 'Política');
    content = content.replaceAll('día', 'día');
    content = content.replaceAll('Día', 'Día');
    content = content.replaceAll('guía', 'guía');
    content = content.replaceAll('Guía', 'Guía');
    content = content.replaceAll('vacío', 'vacío');
    content = content.replaceAll('río', 'río');
    content = content.replaceAll('mínimo', 'mínimo');
    content = content.replaceAll('límite', 'límite');
    content = content.replaceAll('Límite', 'Límite');
    
    content = content.replaceAll('í', 'í');
    
    if (content != original) {
      file.writeAsStringSync(content);
      print('Fixed: \${file.path}');
    }
  } catch (e) {
    // skip non-utf8 files
  }
}

void main() {
  final dir = Directory('.');
  for (final entity in dir.listSync(recursive: true)) {
    if (entity is File) {
      if (entity.path.contains('.git') || 
          entity.path.contains('.dart_tool') || 
          entity.path.contains('build') || 
          entity.path.contains('node_modules') || 
          entity.path.contains('.venv') || 
          entity.path.contains('__pycache__')) {
        continue;
      }
      
      if (entity.path.endsWith('.dart') || 
          entity.path.endsWith('.py') || 
          entity.path.endsWith('.md')) {
        fixMojibake(entity);
      }
    }
  }
  print('Done fixing mojibake.');
}
